### ISSUE TITLE
[Bug] Block settings changes are discarded after uploading a video and saving post

### ISSUE THREAD

**Issue by fluiddot** [wordpress-mobile/gutenberg-mobile#5697]
**Internal reference:** p5T066-3Xm-p2#comment-14605

**Fixes:**
- [x] iOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/20606
- [x] Android: - https://github.com/wordpress-mobile/WordPress-Android/pull/18327

**Comment by fluiddot** [comment]
I originally thought that it might be a race condition between updating the metadata and syncing block settings. However, I debugged the issue and seems the POST request finishes successfully before getting the metadata after saving 🙃. Even though the request responds successfully, the metadata is not updated. I need to investigate further what's happening in the backend that prevents the data to be updated.

**Comment by fluiddot** [comment]
I debugged the API requests on iOS with a proxy and found the culprit. After uploading a video, saving a post executes the following requests:
### 1 - VideoPress metadata update
- Update `description` to `Example`. This request is triggered by the `useSyncMedia` when saving the post. **At this point, the metadata is correct.**
- `POST`: `wpcom/v2/sites/<SITE_ID>/videopress/meta?_locale=user&locale=en`
- Request data
```
{
	"description": "Example",
	"id": <MEDIA_ID>
}
```
- Response
```
{
	"code": "success",
	"message": "Video meta updated successfully.",
	"data": 200
}
```

### 2 - Media update
- Update the metadata of the media item. This request is triggered when saving a post and only after the upload has finished. ❌ **This unexpected request seems the culprit of the issue, as it's overwriting the metadata with old values.**
- `POST`: `rest/v1.1/sites/<SITE_ID>/media/<MEDIA_ID>?locale=en`
- Request data
```
{
	"alt": "",
	"description": "",  ====> THIS VALUE SHOULD BE "Example"
	"caption": "",
	"title": "demo-short-1",
	"parent_id" <POST_ID>
}
```
- Response
```
{
	"ID": <MEDIA_ID>,
	"post_ID": <POST_ID>,
	"title": "demo-short-1",
	"caption": "",
	"description": "",
	"alt": "",
	"videopress_guid": "<VIDEOPRESS_ID>",
	...
}
```

**Comment by fluiddot** [comment]
I checked on iOS and seems the media update is expected for uploads that have been added to the post ([reference](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Services/PostService%2BUnattachedMedia.swift)). When saving a post, we check media items that are unattached to the post (i.e. they don't have the `postID` defined) and update them. This conflicts with the metadata update triggered by `emitPostSaveEvent` as it happens before the media are updated (as shared in [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697#issuecomment-1521421691)).

The potential workarounds I see are:
1. Emit the post save event after the media update is done. This implies postponing the event to a later phase of the saving process, once the post is saved. The potential problem with this approach is that postponing the event might not allow the VideoPress block to make the metadata update request on time before the editor is destroyed.
2. The update related to unattached media items is meant to only update the `postID` field, however, it's updating other fields like `description` or `title`. The media update endpoint allows updating specific fields, hence, we could update that logic to only update `postID`. This would avoid this update to overwrite/discard the metadata updated by the VideoPress block.

**Comment by fluiddot** [comment]
> 1. Emit the post save event after the media update is done. This implies postponing the event to a later phase of the saving process, once the post is saved. The potential problem with this approach is that the postpone the event might not allow the VideoPress block to make the metadata update request on time before the editor is destroyed.

Looks like on iOS the post saving can happen asynchronously ([reference](https://github.com/wordpress-mobile/WordPress-iOS/blob/3ab3c01053c7fc17538d2577d7c874ed81c4de62/WordPress/Classes/ViewRelated/Post/PostEditor%2BPublish.swift#L180-L183)), in fact, this behavior is applied when saving the post upon closing the editor. In this case, we won't be able to guarantee that the post save event is emitted after the post is saved, i.e. after the media is updated.

**UPDATE:** I tried to postpone it and although the VideoPress metadata request is made later, the values displayed in the block settings are the old ones 🤔  It's interesting because the metadata is updated properly, i.e. the media update made when saving the post is not discarding them. However, the fetch made after the metadata update is not reflecting the changes or it's getting the values made by the former media update 🙃 .

**Comment by fluiddot** [comment]
I checked the issue on **Android** and couldn't reproduce it. I was curious about the difference between platforms, especially regarding emitting the post save event, and noticed that the flow is quite similar. The VideoPress metadata is updated before the post is saved and media update is performed after that. However, the difference on Android is that empty values are not included in the `POST` request when updating the media ([reference](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5805ef65e9cebb0153c38ebf98a8f6fab646648c/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java#L621-L640)) 😅 . Following the values passed in the request ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697#issuecomment-1521421691)), the empty ones like `description` are not updated so the metadata updated by the block is preserved.

**Nevertheless, the issue can be reproduced with non-empty values like `title`, so the same issue happens on Android but only when updating `title`.**

**Comment by fluiddot** [comment]
> Looks like on iOS the post saving can happen asynchronously ([reference](https://github.com/wordpress-mobile/WordPress-iOS/blob/3ab3c01053c7fc17538d2577d7c874ed81c4de62/WordPress/Classes/ViewRelated/Post/PostEditor%2BPublish.swift#L180-L183)), in fact, this behavior is applied when saving the post upon closing the editor. In this case, we won't be able to guarantee that the post save event is emitted after the post is saved, i.e. after the media is updated.

As I suspected, postponing emitting the post save event for the async case is not properly working. Although the event is triggered after calling the save post function, the VideoPress metadata request is made before the media update. I also tried to trigger the event on the completion of the async case, but this results on the VideoPress metadata request not being made (most likely because the editor is already destroyed when triggering the event).

Being this said, and at least for iOS, the first option (_Emit the post save event after the media update is done_) is not reliable 😭 .

**Comment by fluiddot** [comment]
After discussing internally (p1682436005524989/1682327360.709399-slack-C04LWMHSGLC), we decided to go with option 2 (Only update `postID` field when updating unattached media items).

**Comment by fluiddot** [comment]
Closing the issue as fixes for both platforms have been merged.
